### PR TITLE
Flash manipulation

### PR DESCRIPTION
**Pull request by stople**
This PR adds functionality to read flash, and manipulate bootloader area of flash.

Ordinary users will likely not need the feature of writing to bootloader area of flash every day. Maybe they will only need it once, ever, when they replace the bad bootloader 2 with a good bootloader 2 (or a newer bootloader).

However. There is a risk involved in supporting commands which can write to the bootloader flash, as bootloader can be unintentional (or even malicious) manipulated. Thus, I suggest to at least protect this feature by not enabling it by default, it requires a specially built version which includes this feature. This also frees up some space for other features. Additionally, we could also add some button protection for the feature, I am testing this in branch "make-programming-safer".

Let me know what you think:
- Should we require a special FW to use the flash manipulation features (to avoid unintentional use), or should we include this functionality in standard SMC firmware?
- If we only include the flash manipulation feature in a special SMC FW, should the feature be button-protected?
- Do you agree that we should limit flash manipulation functionality to bootloader area only? This will at least ensure that the flash manipulation functionality does not overwrite itself/the running program, reducing the chance of bricking the SMC

### CODE REVIEW

**Comment by stefan-b-jakobsson** [comment]
I think it's reasonable that the firmware can only change the bootloader area, and that it can't update the firmware. It's unlikely that that could work in any safe manner.

I don't think we need to worry too much about malicious attacks on the SMC or the bootloader. But it would be great to protect the firmware from unintentional updates, for instance by requiring a button press.

It would be nice to keep the bootloader update functionality in the firmware. We can't rule out that we will need this functionality in the future as well :-)

**Comment by stefan-b-jakobsson** [comment]
To save space, maybe the button-protect could be commonly used both for the update of the bootloader area and for starting the bootloader.

**Comment by stople** [comment]
> It would be nice to keep the bootloader update functionality in the firmware. We can't rule out that we will need this functionality in the future as well :-)

I agree that we may need this again, once bootloader version 3 is ready. However, it is possible to have a special purpose SMC version, which can be installed first in order to patch the bootloader, and then you install the ordinary SMC firmware which is not built with this feature. This is one option, to avoid unintentional use of this feature. But another option is off course to include it permanently, and only protect it with a button requirement. I guess it boils down to convenience vs reducing the risk of bricking. But if you think convenience is better in this case, I will add the button commits to this PR and some matching scripts to make use of the feature.

**Comment by jburks** [comment]
I have a somewhat crazy idea: is it possible to wedge this into the firmware with the bad bootloader such that page 0 is not overwritten when updating with the bad bootloader? I'm thinking on a way to get the bad bootloader fixed on the hundreds of machines that have already shipped with it. Perhaps take an existing I2C command (maybe the echo command) and have it jump to this new code so a sequence for those affected could be:

1. Run a special update program that patches the bad SMC firmware with the bootloader flash update code
2. Reset
3. Run another special update program that patches the bad bootloader
4. Run the standard flash program to update the SMC

**Comment by stople** [comment]
> I have a somewhat crazy idea: is it possible to wedge this into the firmware with the bad bootloader such that page 0 is not overwritten when updating with the bad bootloader? I'm thinking on a way to get the bad bootloader fixed on the hundreds of machines that have already shipped with it. Perhaps take an existing I2C command (maybe the echo command) and have it jump to this new code so a sequence for those affected could be:
> 
> 1. Run a special update program that patches the bad SMC firmware with the bootloader flash update code
> 2. Reset
> 3. Run another special update program that patches the bad bootloader
> 4. Run the standard flash program to update the SMC

Not sure if I completely follow, but my idea is as follows:
- Use the regular update program to update to a SMC FW version which includes this flash manipulation functionality (either a special SMC FW version for this purpose, or, as a feature in standard SMC FW). Users which have the bad bootloader will have to manually poke the reset pin, to avoid zero page to be filled with random data.
- Use a special program which uses these new features, to check exactly which bootloader is installed (by dumping its flash), and optionally replace the bad bootloader with either the correct v2 bootloader or a future v3 bootloader (or even roll back to the old v1 bootloader if you like, I have experimented with this and it works)

**Comment by stople** [comment]
Maybe we should also modify the regular update program, so that it does NOT suggest users of "bootloader v2" to unplug power when done. If the user read this text, there is a big risk that they will do as they are told, and unplug the power, causing a bricked SMC. The updated program should mention something like "if you can read this text with a bootloader v2, it means like you likely have a known, but corrupt, version of the v2 bootloader. DO NOT unplug power, as that will brick your SMC. Instead, complete the programming by manually shorting the SMC reset pin to ground, ref update instructions.". Perhaps there should also be added a new warning page if bootloader is version 2, so that the user can know before programming is done, that he have a version which MAY be the bad one, so that he is prepared according to the update instructions (e.g. have a jumper wire ready and know what to do with it if needed).

**Comment by stople** [comment]
[smc-flash-manipulation-2.zip](https://github.com/X16Community/x16-smc/files/15288006/smc-flash-manipulation-2.zip)

Here are some tools to read and write flash, compatible with this PR (including button protected activation).

I see that there are no other I2C commands that are both read and write. Should I change command $93 so that there are unique numbers for read and write?

File: README.md
Comment (stefan-b-jakobsson) [inline README.md:44]: I think it's OK to use 0x93 as an offset that you can both read and write.

File: x16-smc.ino
Comment (stefan-b-jakobsson) [inline x16-smc.ino:88]: "Special Mode" is not very descriptive.

"Self Programming Mode"?

File: x16-smc.ino
Comment (stefan-b-jakobsson) [inline x16-smc.ino:274]: This if statement does nothing and is kind of obsolete. I think it should be removed.

File: x16-smc.ino
Comment (stefan-b-jakobsson) [inline x16-smc.ino:294]: This if statement does nothing. I think it should be removed.

File: x16-smc.ino
Comment (stefan-b-jakobsson) [inline x16-smc.ino:312]: Same. Obsolete if statement

File: x16-smc.ino
Comment (stefan-b-jakobsson) [inline x16-smc.ino:469]: Power+NMI commented below

File: x16-smc.ino
Comment (stefan-b-jakobsson) [inline x16-smc.ino:129]: selfProgrammingModeRequested?

File: x16-smc.ino
Comment (stefan-b-jakobsson) [inline x16-smc.ino:130]: selfProgrammingModeActive?

File: x16-smc.ino
Comment (stefan-b-jakobsson) [inline x16-smc.ino:649]: I'm not sure I follow the exact logic of disabling interrupts during this sequence of code.

I understand that startBootloader is called from the main loop, and could be interrupted.

The value of bootloaderTimer could be changed within in an interrupt before the remainder of this function has been executed.

File: x16-smc.ino
Comment (stefan-b-jakobsson) [inline x16-smc.ino:663]: Note that not all variants of the X16 board have an NMI button. And a lot of cases only have two buttons. As the NMI button is the least useful of the three buttons, it's like not connected. Would it be more convenient to use Power+Reset for both modes?

**Comment by stefan-b-jakobsson** [comment]
I need to test the bootloader update utility a bit more, but it looks like a success.

There is a lot of overhead in the I2C transmission set up like this, but the bootloader is so tiny it doesn't really matter.

File: x16-smc.ino
Comment (stople) [reply x16-smc.ino:274]: It does serve a purpose as is, which is to "disable the power button" while bootloaderTimer is active (as it then prevents the following "else if" blocks to execute). You do not want the SMC to actually power off the power supply when doing the button combination.

I can improve the comment to highlight this effect. Or, I can replace it with a "return" if that makes it cleaner. Or, do you see any other solution, as you suggest to remove it?

The same applies to reset and NMI. I opted for using the same logic all places.

File: x16-smc.ino
Comment (stople) [reply x16-smc.ino:670]: Good point. I agree that we should make it usable on machines without a dedicated NMI button if that is a thing. My thought was to prevent starting this feature accidentally (or maliciously, maybe I am a bit too paranoid). I can try to simply change buttons to power+reset on this mode as well.

How does this require you to hold buttons longer? Have you experienced this, or is it based on something you read in the code?

Comment (stefan-b-jakobsson) [reply x16-smc.ino:670]: It is a bit different. The current code doesn't really require both buttons to be pressed at the same time. It only requires them to be pressed/released within the 0.5 s threshold. I managed to press/release them in a way that the new code didn't accept it.

I don't know if it needs to be changed. It's more of an observation.

File: x16-smc.ino
Comment (stefan-b-jakobsson) [reply x16-smc.ino:274]: It's just a bit strange to have an if statement that does nothing.

Maybe it's best to keep it. It could be rewritten like this though:

```
if (bootloaderTimer == 0) {
  if (SYSTEM_POWERED == 0) {                  // If Off, turn on
    PowerOnSeq();
  }
  else {                                      // If On, turn off
    PowerOffSeq();
  }
}
```

File: x16-smc.ino
Comment (stople) [reply x16-smc.ino:649]: It is because the 16 bit volatile variable bootloaderTimer is decremented in main loop while it can be set elsewhere (likely in interrupts). Given the code

```
    bootloaderTimer--;
    15ca:	80 91 53 01 	lds	r24, 0x0153	; 0x800153 <bootloaderTimer>
    15ce:	90 91 54 01 	lds	r25, 0x0154	; 0x800154 <bootloaderTimer+0x1>
    15d2:	01 97       	sbiw	r24, 0x01	; 1
    15d4:	90 93 54 01 	sts	0x0154, r25	; 0x800154 <bootloaderTimer+0x1>
    15d8:	80 93 53 01 	sts	0x0153, r24	; 0x800153 <bootloaderTimer>
```

, 15d4 writes the high value back to ram. 15d8 writes the low value back to ram. There can trigger an interrupt between 15d4 and 15d8 changing both places in ram. When 15d8 executes after the interrupt, the value of bootloaderTimer will not contain an intended value. To ensure that the value of bootloaderTimer is always an intended value, I do the decrement as an atomic operation. (Perhaps I should use ATOMIC_BLOCK_RESTORE instead to make the intent more clear). Also, there can trigger an interrupt between 15ca and 15ce, also causing that wrong number gets decremented.

Comment (stefan-b-jakobsson) [reply x16-smc.ino:649]: OK, that makes sense.

**Comment by stefan-b-jakobsson** [comment]
I've done some testing without having found any errors:

- Writing to page 0 and 119 (firmware area) is ignored
- Writing to page 120 to 127 (bootloader area) works
- Setting page to 128 wraps around to page 0. Writing is ignored.
- If you first enable the self-programming mode, and then disable it again, writing to the flash is ignored
- If you change the flash page before writing 64 bytes, the bytes so far written are lost, and you start over from byte 0
- Reading from all parts of the flash memory works
- Updating the bootloader works, and the bootloader runs properly

I think this PR should be accepted, but you should first consider the above mentioned changes:
- The name of the special mode
- Using Power+Reset to enter self-programming mode both for the firmware area and the bootloader area.

**Comment by stople** [comment]
I have now updated this PR.

Changes:
- Button combination logic have been reverted to the original (short press within the same time period)
- Buttons have been changed to Power+Reset, same buttons as used to start bootloader. Rationale is that some X16s may omit the NMI button.
- Names of variables and enums have been changed according to PR feedback

Updated script attached, which mentions the updated button combination.

[smc-flash-manipulation-3.zip](https://github.com/X16Community/x16-smc/files/15409318/smc-flash-manipulation-3.zip)

I have tested dumping the bootloader, and changing bootloader version using the button combination. I have also tested the button combination while starting the bootloader.

**Comment by stefan-b-jakobsson** [comment]
I’m away from home a few days and will not be able to test until beginning of next week.

**Comment by stople** [comment]
@stefan-b-jakobsson @jburks @mooinglemur FYI, here is a preliminary "recovery bundle" explaining how to update PR boards with bad bootloader (at least PR00100 to PR00831 with SMC version 45.1.0) to a newer SMC version. Preferably a future SMC version 48.x.x, including this PR. Then, the user can use the included tools to fix the bootloader. Posting it here first, to get your feedback, before posting it on Discord (as a prelim). This should preferably be tested by someone able to recover the SMC in case of failure. 
[smc recovery bundle PRELIM.zip](https://github.com/X16Community/x16-smc/files/15506197/smc.recovery.bundle.PRELIM.zip)
To test it, start with this file "x16-smc-r45.1-bootloader.hex" with bad v2 bootloader, found here https://discord.com/channels/547559626024157184/1224366756320247942/1225169489662836908

**Comment by stefan-b-jakobsson** [comment]
I read through the instructions, and I can't see any other way to do this.

Should we recommend people using a test clip instead of touching the reset pin freehand? Example product: https://www.digikey.com/en/products/detail/digilent-inc/240-137/9916326

**Comment by stople** [comment]
I was thinking that most users have, or can borrow, a jumper wire. I doubt that most users have a test clip available, they will then have to order something. If they anyway order something, they can also purchase a real programmer (although a test clip is likely a bit cheaper). But, given that they do have a test clip at hand. What do you recommend them to do? Do you have a test clip at hand? Can you try to follow the guide with a test clip, and give some feedback (with some more details) if you think the test clip approach adds some value to the instructions?

As for the neighbor pins, it seems like we are lucky. Pin 9 and 8 is PS2 mouse clock and data, which is inputs with a weak (too weak :P ) pullup. Pin 7 is reset button input, which is active low with internal pullup. Pin 6 is ground. Pin 5 is VCC. So, as long as the user stays away from pin 5, it seems like it should not damage anything.

```
/*  - Pinout Updated for Proto 4 / Dev Board
ATTINY861 Pinout
     AVR Func         X16 Func   ArdIO   Port             Port   ArdIO   X16 Func    AVR Function
                                              ----\_/----
                                             | *         |
 (SPI MOSI) (SDA)      I2C_SDA     8     PB0 | 1       20| PA0     0     RESB
 (SPI MISO)               IRQB     9     PB1 | 2   A   19| PA1     1     NMIB
  (SPI SCK) (SCL)      I2C_SCL    10     PB2 | 3   T   18| PA2     2     PS2_KBD_CLK
                   PS2_KBD_DAT    11     PB3 | 4   t   17| PA3     3     POWER_OK
                                         VCC | 5   i   16| AGND
                                         GND | 6   n   15| AVCC
                     RESET_BTN    12     PB4 | 7   y   14| PA4     4     POWER_BTN
                   PS2_MSE_DAT    13     PB5 | 8   8   13| PA5     5     POWER_ON
                   PS2_MSE_CLK    14     PB6 | 9   6   12| PA6     6     ACT_LED        (TXD)
  (SPI SS) (RST)                  15     PB7 |10   1   11| PA7     7     NMI_BTN        (RXD)            
                                             |           |
                                              -----------
 */
```

**Comment by stefan-b-jakobsson** [comment]
OK, if we can conclude that the risk of damaging anything by grounding the nearby pins is low I have no objection against recommending end users to use a jumper wire.
If I find time this weekend, I will test the whole upgrade procedure. From my earlier tests, I'm pretty sure it will work.

**Comment by stefan-b-jakobsson** [comment]
I've tested the SMC upgrade procedure, and it worked fine.

Some notes on my test:
- Hardware: OtterX board
- Firmware update tool: https://github.com/stefan-b-jakobsson/x16-smc-update
- First, I installed the bad bootloader from the Discord link referred to above
- Second, I run the firmware update tool to verify that it fails. The firmware update program returned without errors, but the system didn't reset. After disconnecting and reconnecting mains power, the system was dead, as expected.
- Third, I reinstalled the bad bootloader.
- Fourth, I downloaded the last SMC artifact for this PR, https://github.com/X16Community/x16-smc/actions/runs/9199275352
- Fifth, I used the firmware update tool to install the downloaded hex file. As before, the update tool exited but didn't reset. I used a jumper wire to connect the ATtiny reset pin to ground. The system immediately shut down. I could then turn it on again by pressing the power button. Everything worked OK as far as I could tell.
- Sixth, I run SMCBLW16.PRG to install a new bootloader. That worked without errors.
- Seventh, I finally run the firmware update tool again. This time, the system reset itself when the update was done, showing that the bootloader had indeed been fixed.

In summary, I think we have a winner.

We might want to work some more on the documentation though.

**Comment by stople** [comment]
Ref Discord, https://discord.com/channels/547559626024157184/1248410163468046336 , two users (Stefan and squarewave) have succeeded in updating the bootloader using the changes in this PR.

I agree that documentation may be improved. But, is that required prior to merging this PR? Do you want that documentation inside the repo somehow? Because, I suggest to give a new version number to SMC after this merge, so that we, in the documentation, can refer to this version as "minimum version" for supporting reading and writing of flash.

This PR is ready from my side. Let me know if you expect me to do something more before this can be merged. Waiting for review and eventually instructions for how to bump version number. @stefan-b-jakobsson @jburks

**Comment by stefan-b-jakobsson** [comment]
Yes, I think this PR should be merged.
Unless someone opposes it, I can do it this weekend.
The version number is set in version.h. There's nothing more to it than that.

**Review by stefan-b-jakobsson — APPROVED** [review]

**Pull request merged**
